src/lexer.py

```diff
@@ -290,6 +339,136 @@ def word_symbolizer(source: str, in_idx: int, error_generator: ErrorGenerator) -
+        idx = skip_whitespace(source, idx)
+        if source[idx] != "\n":
+            raise datatypes.LexerError(error_message, location=location, start_idx=start_idx, end_idx=idx)
+        return True, idx, datatypes.Token(type=datatypes.TokenType.Number, data=number, location=location)
+
+    # If the start looks like a string, assume we're matching a string
+    if current_char == '"':
+        idx, string = parse_string(source, idx, error_func)
+        location = get_srcloc(line, line_start, start_idx, idx)
+        idx = skip_whitespace(source, idx)
+        if source[idx] != "\n":
+            raise datatypes.LexerError(error_message, location=location, start_idx=start_idx, end_idx=idx)
+        return True, idx, datatypes.Token(type=datatypes.TokenType.String, data=string, location=location)
+
+    idx, first_word = get_next_word(source, idx)
+    match: Optional[datatypes.TokenType] = SINGLE_KEYWORDS.get(first_word)
+    # If this is a poetic type literal assignment, just match that
+    if match is not None and match in TYPE_CONSTANT_SYMBOLS:
+        location = get_srcloc(line, line_start, start_idx, idx)
+        contents = source[start_idx:idx]
+        idx = skip_whitespace(source, idx)
+        if source[idx] != "\n":
+            raise datatypes.LexerError(error_message, location=location, start_idx=start_idx, end_idx=idx)
+        return True, idx, datatypes.Token(type=match, data=contents, location=location)
+    return False, -1, None
+
+def tokenize_poetic_assignment(source: str, line: int, line_start: int, start_idx: int) -> Tuple[int, datatypes.Token]:
+    """
+    Assuming we've started a poetic type literal or poetic number assignment, attempt to tokenize
+    the remainder of the line as a type literal or poetic number. It is an error for a
+    poetic type literal assignment to include words after the type literal.
+
+    :param source:                   String of the source file
+    :param line:                     Index of the line in the source file, used to construct location
+    :param line_start:               Index of the first character in this line, used to construct location
+    :param start_index:              The index in the source file to start tokenizing from
+    :return:                         Tuple containing the next index to tokenize from and the token produced
+    :raises datatypes.LexerError:    On text that looked like it was a number or string but failed to parse correctly,
+                                     as well as when extra text follows the poetic type literal.
+    """
+    # At this point we don't know if this is a poetic type literal assignment or a
+    # poetic number assignment. To figure it out, we try to tokenize the first word
+    # as a type literal. If it fails, it must be a poetic number assignment.
+    success, result_idx, result_token = try_type_literal_assignment(source, line, line_start, start_idx)
+    if success:
+        assert result_token is not None
```

Comment:
See return type comment above. This is an invariant that can be assured by a refactor, making the design less brittle.